### DISCUSSION TITLE
Serial2 shell commands not working

### DISCUSSION THREAD

**Discussion by feraloid** [OpenGammaProject/Open-Gamma-Detector#52] [Q&A]
While connecting the OpenGamma Detector to an ESP32 via the UART/Serial2 pins (H1 #9 & 10) I have run into a problem. The documentation seems to say the OGD will accept shell commands via the UART serial (Serial2) port. The data transmitted from the OGD is coming through fine on the ESP32. 

When I attempt to transmit shell commands from the ESP32 to the OGD, they are apparently ignored. I've mirrored the characters typed onto the ESP32 in code via Bluetooth serial to make sure the typed characters are being received/processed by the ESP32 code. I've also switched to lower baud rates on the chance there's a weird baud rate issue.  I have tried different transmit pins off the ESP32 to make sure I didn't have a bad pin. Finally, just for completeness sake, I checked my transmit pin on the ESP32 with an oscilloscope and verified there is data coming through.

Did I misunderstand that the OGD will accept shell commands from the UART port or do I have some weird problem I have failed to find?

**Comment by NuclearPhoenixx** [comment]
You're right, thanks for letting me know. It didn't work because of the Serial shell library that I'm using was using the USBSerial only. I fixed that bug and now both serial ports should be working as expected with firmware version 4.2.1. Please let me know if it works for you now!

**Reply by feraloid** [reply]
Now I'm getting an error during compile:
C:\Users\gbrus\Documents\Arduino\ogd_pico\ogd_pico.ino:147:20: error: no matching function for call to 'CShell::CShell(SerialUSB&)'
  147 | CShell shell(Serial);
      |                    ^
In file included from C:\Users\gbrus\Documents\Arduino\ogd_pico\ogd_pico.ino:34:
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note: candidate: 'CShell::CShell()'
   33 | class CShell {
      |       ^~~~~~
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note:   candidate expects 0 arguments, 1 provided
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note: candidate: 'CShell::CShell(const CShell&)'
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note:   no known conversion for argument 1 from 'SerialUSB' to 'const CShell&'
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note: candidate: 'CShell::CShell(CShell&&)'
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note:   no known conversion for argument 1 from 'SerialUSB' to 'CShell&&'
C:\Users\gbrus\Documents\Arduino\ogd_pico\ogd_pico.ino:148:22: error: no matching function for call to 'CShell::CShell(SerialUART&)'
  148 | CShell shell1(Serial2);
      |                      ^
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note: candidate: 'CShell::CShell()'
   33 | class CShell {
      |       ^~~~~~
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note:   candidate expects 0 arguments, 1 provided
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note: candidate: 'CShell::CShell(const CShell&)'
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note:   no known conversion for argument 1 from 'SerialUART' to 'const CShell&'
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note: candidate: 'CShell::CShell(CShell&&)'
c:\Users\gbrus\Documents\Arduino\libraries\SimpleShell_Enhanced\src/SimpleShell_Enhanced.h:33:7: note:   no known conversion for argument 1 from 'SerialUART' to 'CShell&&'

exit status 1

Compilation error: no matching function for call to 'CShell::CShell(SerialUSB&)'

**Reply by NuclearPhoenixx** [reply]
Yeah I know, the Arduino library update is not through yet. It takes hours for them to get the updates into the IDE, unfortunately. Please use the `.uf2` files for now.

**Comment by feraloid** [comment]
That did the trick! Serial2/UART shell commands are working now. I'm relieved that I hadn't screwed up something horribly (at least this time). Thanks so much!
